proxy/src/main/java/net/md_5/bungee/util/BungeeTranslationUtil.java

```diff
@@ -0,0 +1,146 @@
+package net.md_5.bungee.util;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.chat.ComponentBuilder;
+import net.md_5.bungee.api.chat.TextComponent;
+import net.md_5.bungee.api.chat.TranslatableComponent;
+import org.jetbrains.annotations.NotNull;
+
+public class BungeeTranslationUtil
+{
+    // support only simple translations, no escaping and ignore specialized formatting
+    private static final Pattern TRANSLATE_FORMAT = Pattern.compile( "\\{(\\d*)(?:,[^}]+)?}" );
+
+    public static @NotNull BaseComponent getTranslationComponent0(String trans, Object... args)
+    {
+
+        Matcher matcher = TRANSLATE_FORMAT.matcher( trans );
+        int position = 0;
+        int i = 0;
+        BaseComponent lastNonArg = null;
+        ComponentBuilder builder = new ComponentBuilder();
+        while ( matcher.find() )
+        {
+            // Add the static part of the string
+            int pos = matcher.start();
+            if ( pos != position )
+            {
+                builder.appendLegacy( trans.substring( position, pos ) );
+                BaseComponent curr = builder.getCurrentComponent();
+                if ( lastNonArg != null )
+                {
+                    // undo eventual formatting of previous argument
+                    boolean colorChange = curr.isReset();
+                    if ( !Objects.equals( lastNonArg.getColor(), curr.getColor() ) )
+                    {
+                        colorChange = true;
+                        builder.color( lastNonArg.getColor() );
+                    }
+                    if ( colorChange && lastNonArg.isBold() || lastNonArg.isBold() != curr.isBold() )
+                    {
+                        builder.bold( lastNonArg.isBold() );
+                    }
+                    if ( colorChange && lastNonArg.isItalic() || lastNonArg.isItalic() != curr.isItalic() )
+                    {
+                        builder.italic( lastNonArg.isItalic() );
+                    }
+                    if ( colorChange && lastNonArg.isUnderlined() || lastNonArg.isUnderlined() != curr.isUnderlined() )
+                    {
+                        builder.underlined( lastNonArg.isUnderlined() );
+                    }
+                    if ( colorChange && lastNonArg.isStrikethrough() || lastNonArg.isStrikethrough() != curr.isStrikethrough() )
+                    {
+                        builder.strikethrough( lastNonArg.isStrikethrough() );
+                    }
+                    if ( colorChange && lastNonArg.isObfuscated() || lastNonArg.isObfuscated() != curr.isObfuscated() )
+                    {
+                        builder.obfuscated( lastNonArg.isObfuscated() );
+                    }
+                }
+                lastNonArg = findLastComponent( builder.getCurrentComponent().duplicate() );
```

Comment:
Doesn't it make more sense to duplicate the result of findLast instead?

Reply:
That would not retain the parents of the component (duplicate is recursive), but with ComponentStyle, we now have a handy object to save the effective formatting of lastNonArg  to and get rid of duplicating at all.